controllers/upgrade_controller.go

```diff
@@ -0,0 +1,270 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+const (
+	ocsClientOperatorSubscriptionPackageName = "ocs-client-operator"
+	// TODO(lgangava): there can be "eus" or "fast" or "candidate" supported channels
+	// 	and custom channels also might exist, either get this data as part of a configmap
+	// 	or cut the name by '-'
+	subscriptionChannelNamePrefix = "stable-"
+	controllerName                = "upgrade"
+	upgradeConditionReason        = "ConstraintsSatisfied"
+	dummyPatchVersion             = ".99"
+)
+
+type UpgradeReconciler struct {
+	client.Client
+	ctx context.Context
+	log logr.Logger
+
+	Scheme            *runtime.Scheme
+	OperatorCondition conditions.Condition
+	OperatorNamespace string
+	PlatformVersion   string
+}
+
+func (r *UpgradeReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	storageClientOperatorVersionPredicate := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if e.ObjectOld == nil || e.ObjectNew == nil {
+				return false
+			}
+
+			oldObj, _ := e.ObjectOld.(*v1alpha1.StorageClient)
+			newObj, _ := e.ObjectNew.(*v1alpha1.StorageClient)
+			if oldObj == nil || newObj == nil {
+				return false
+			}
+
+			if newObj.Status.Operator.CurrentVersion != newObj.Status.Operator.DesiredVersion {
+				return true
+			}
+
+			return false
+		},
+	}
+	ocsClientOperatorSubscriptionPredicate := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if e.ObjectOld == nil || e.ObjectNew == nil {
+				return false
+			}
+
+			oldObj, _ := e.ObjectOld.(*opv1a1.Subscription)
+			newObj, _ := e.ObjectNew.(*opv1a1.Subscription)
+			if oldObj == nil || newObj == nil {
+				return false
+			}
+
+			if oldObj.Spec.Package != ocsClientOperatorSubscriptionPackageName {
+				return false
+			}
+
+			return oldObj.Spec.Channel != newObj.Spec.Channel
+		},
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(controllerName).
+		Watches(&v1alpha1.StorageClient{}, &handler.EnqueueRequestForObject{}, builder.WithPredicates(storageClientOperatorVersionPredicate)).
+		Watches(&opv1a1.Subscription{}, &handler.EnqueueRequestForObject{}, builder.WithPredicates(ocsClientOperatorSubscriptionPredicate)).
+		Complete(r)
+}
+
+//+kubebuilder:rbac:groups=operators.coreos.com,resources=subscriptions,verbs=get;list;patch
+
+func (r *UpgradeReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	r.log = log.FromContext(ctx)
+	r.log.Info("starting reconcile")
+
+	conditionStatus := metav1.ConditionTrue
+	conditionMessage := "No component reported errors."
+	result, err := r.reconcilePhases()
+	if err != nil {
+		// TODO (lgangava): we are being very restrictive as of now, not upgrading if we hit any error
+		// differentiate between reconciler errors and not upgradeable errors
```

Comment:
Don't those coincide? Meaning surely there are some reconciling errors that imply non-upgradability

Reply:
> there are some reconciling errors that imply non-upgradability
- currently we are implying all reconcile errors as non-upgradability
- do you mean to say, not all reconcile errors shouldn't be viewed as not upgradable?